lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -631,6 +644,421 @@ int getBranchCount(Value val, Block *block) {
+  auto externalCtrlMerge = rewriter->create<ControlMergeOp>(loc, externalCtrls);
+
+  // Create loop mux and the loop priming register. The loop mux will on select
+  // "0" select external control, and internal control at "1". This convention
+  // must be followed by the loop exit network.
+  auto primingRegister =
+      rewriter->create<BufferOp>(loc, rewriter->getI1Type(), /*size=*/1,
+                                 loopPrimingInput, /*sequential=*/true);
+  // Initialize the priming register to path 0.
+  primingRegister->setAttr("initValues", rewriter->getI64ArrayAttr({0}));
+
+  // The loop control mux will deterministically select between control entering
+  // the loop from any external block or the single loop backedge.
+  auto loopCtrlMux = rewriter->create<MuxOp>(
+      loc, primingRegister.getResult(),
+      llvm::SmallVector<Value>{externalCtrlMerge.result(), loopCtrl});
+
+  // Replace the existing control merge 'result' output with the loop control
+  // mux.
+  cmerge->getResult(0).replaceAllUsesWith(loopCtrlMux.getResult());
+
+  // Register the new block entry control value
+  fol.setBlockEntryControl(loopHeader, loopCtrlMux.getResult());
+
+  // Next, we need to consider how to replace the control merge 'index' output,
+  // used to drive input selection to the block.
+
+  // Inputs to the loop header will be sourced from muxes with inputs from both
+  // the loop latch as well as external blocks. Iterate through these and sort
+  // based on the input ordering to the external/internal control merge.
+  // We do this by maintaining a mapping between the external and loop data
+  // inputs for each data mux in the design. The key of these maps is the
+  // original mux (that is to be replaced).
+  DenseMap<MuxOp, std::vector<Value>> externalDataInputs;
+  DenseMap<MuxOp, Value> loopDataInputs;
+  for (auto muxOp : muxesToReplace) {
+    if (muxOp == loopCtrlMux)
+      continue;
+
+    externalDataInputs[muxOp] = getSortedInputs(externalCtrlMerge, muxOp);
+    loopDataInputs[muxOp] = getOperandFromBlock(muxOp, loopLatch);
+    assert(/*loop latch input*/ 1 + externalDataInputs[muxOp].size() ==
+               muxOp.dataOperands().size() &&
+           "Expected all mux operands to be partitioned between loop and "
+           "external data inputs");
+  }
+
+  // With this, we now replace each of the data input muxes in the loop header.
+  // We instantiate a single mux driven by the external control merge.
+  // This, as well as the corresponding data input coming from within the single
+  // loop latch, will then be selected between by a 3rd mux, based on the
+  // priming register.
+  for (MuxOp mux : muxesToReplace) {
+    auto externalDataMux = rewriter->create<MuxOp>(
+        loc, externalCtrlMerge.index(), externalDataInputs[mux]);
+
+    rewriter->replaceOp(
+        mux, rewriter
+                 ->create<MuxOp>(loc, primingRegister,
+                                 llvm::SmallVector<Value>{externalDataMux,
```

Comment:
nit: I think you could remove `llvm::SmallVector<Value>` here to save some typing.

Reply:
The `ValueRange` doesn't seem to be able to implicitly instantiate just using an initializer list.

Reply:
Hmm gotcha I thought I'd done that before.